### PR TITLE
(4.x) Merge 3.4

### PR DESCRIPTION
**Pull request by alalek**
#21306 from aBITnav:patch-1
#21320 from catree:solvePnP_doc_page
#21321 from alalek:dnn_test_decompose_dynamic_axes
#21322 from alalek:dnn_catch_errors
#21323 from alalek:dnn_do_not_rebuilt_network_in_setInput
#21325 from APrigarina:fix_asift_sample
#21327 from alalek:imgcodecs_openexr_runtime_disabled
#21329 from alalek:cmake_openexr_find_package
#21330 from alalek:fix_highgui_gtk_mouse_callbacks
#21331 from alalek:issue_21289
#21335 from alalek:dnn_onnx_dump_input_output
#21336 from alalek:dnn_pooling_check_array_indexes
#21341 from alalek:issue_20930_3.4

Previous "Merge 3.4": #21317

<cut/>

<details>

```
buildworker:Win64 OpenCL=windows-2
Xbuildworker:Custom=linux-1,linux-2,linux-4,linux-6
buildworker:Docs=linux-4,linux-6
build_image:Docs=docs-js:18.04
build_image:Custom=javascript
buildworker:Custom=linux-4,linux-6
Xbuild_image:Custom=javascript-simd
Xbuild_image:Custom=powerpc64le
Xbuild_image:Custom=ubuntu-openvino-2019r3.0:16.04
Xbuild_image:Custom=ubuntu-openvino-2020.3.0:16.04
Xbuild_image:Custom=ubuntu-openvino-2020.4.0:16.04
Xbuild_image:Custom=ubuntu-openvino-2021.1.0:20.04
Xbuild_image:Custom=ubuntu-openvino-2021.2.0:20.04
Xbuild_image:Custom=ubuntu-openvino-2021.3.0:20.04
Xbuild_image:Custom=ubuntu-openvino-2021.4.0:20.04
Xbuild_image:Custom=ubuntu-openvino-2021.4.1:20.04
Xbuild_image:Custom=ubuntu-openvino-2021.4.2:20.04
Xbuildworker:Custom=linux-1
Xbuild_image:Custom=ubuntu-vulkan:16.04
Xbuildworker:Custom=linux-4
Xbuild_image:Custom=fedora:28
Xbuild_image:Custom=ubuntu-cuda:16.04
Xbuild_image:Custom=ubuntu-clang:18.04
Xbuild_image:Custom=ubuntu:20.04
Xbuildworker:Custom=linux-1
Xbuild_image:Custom=javascript-simd
Xbuild_image:Custom=mips64el

Xbuild_image:Custom Mac=openvino-2019r3.0
Xbuild_image:Custom Mac=openvino-2020.3.0
Xbuild_image:Custom Mac=openvino-2020.4.0
Xbuild_image:Custom Mac=openvino-2021.1.0
Xbuild_image:Custom Mac=openvino-2021.2.0
Xbuild_image:Custom Mac=openvino-2021.3.0
Xbuild_image:Custom Mac=openvino-2021.4.0
Xbuild_image:Custom Mac=openvino-2021.4.1
build_image:Custom Mac=openvino-2021.4.2
test_modules:Custom Mac=dnn,gapi,python2,python3,java

Xbuild_image:Custom Win=openvino-2019r3.0
Xbuild_image:Custom Win=openvino-2020.3.0
Xbuild_image:Custom Win=openvino-2020.4.0
Xbuild_image:Custom Win=openvino-2021.1.0
Xbuild_image:Custom Win=openvino-2021.2.0
Xbuild_image:Custom Win=openvino-2021.3.0
Xbuild_image:Custom Win=openvino-2021.4.0
Xbuild_image:Custom Win=openvino-2021.4.1
build_image:Custom Win=openvino-2021.4.2
buildworker:Custom Win=windows-3
test_bigdata:Custom Win=1
test_filter:Custom Win=*
test_modules:Custom Win=dnn,gapi,python2,python3,java
test_opencl:Custom Win=ON
build_contrib:Custom Win=OFF
Xbuild_image:Custom Win=msvs2017
Xbuild_image:Custom Win=msvs2019
```

</details>

### CODE REVIEW

File: modules/dnn/test/test_onnx_importer.cpp [hunk 1]
```diff
@@ -1344,6 +1489,16 @@ TEST_P(Test_ONNX_layers, DivConst)
...
+// FIXIT disabled due to non-standard ONNX model domains, need to add ONNX domains support
+// Example:
+// DNN/ONNX: unsupported opset[1]: domain='com.microsoft.experimental' version=1
+// DNN/ONNX: unsupported opset[2]: domain='ai.onnx.preview.training' version=1
+// DNN/ONNX: unsupported opset[3]: domain='com.microsoft.nchwc' version=1
+// DNN/ONNX: unsupported opset[4]: domain='com.microsoft.mlfeaturizers' version=1
+// DNN/ONNX: unsupported opset[5]: domain='ai.onnx.ml' version=2
+// DNN/ONNX: unsupported opset[6]: domain='com.microsoft' version=1
+// DNN/ONNX: unsupported opset[7]: domain='ai.onnx.training' version=1
+#if 0
```
Comment (alalek) [inline modules/dnn/test/test_onnx_importer.cpp:1501]: Need to properly implement ONNX domains support for quantized networks.

/cc @vpisarev

**Comment by alalek** [comment]
Face detector (from #20422) was imported through bugs. New checks for unsupported cases emits these errors:

> DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 243 [ 0 0 0 51 ]

```
[ RUN      ] Objdetect_face_detection.regression
[ INFO:0@12.252] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (721) populateNet DNN/ONNX: loading ONNX v6 model produced by 'pytorch':1.9. Number of nodes = 111, initializers = 92, inputs = 1, outputs = 3
[ INFO:0@12.252] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (654) parseOperatorSet DNN/ONNX: ONNX opset version = 9
[ERROR:0@12.256] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2479) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 243 [ 0 0 0 51 ]
[ERROR:0@12.256] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (872) handleNode DNN/ONNX: ERROR during processing node with 1 inputs and 1 outputs: [Shape]:(265)
[ INFO:0@12.256] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (876) handleNode     Input[0] = '243'
[ INFO:0@12.256] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (880) handleNode     Output[0] = '265'
unknown file: Failure
C++ exception with description "OpenCV(4.5.5-pre) /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp:891: error: (-2:Unspecified error) in function 'handleNode'
> Node [Shape]:(265) parse error: OpenCV(4.5.5-pre) /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp:2480: error: (-215:Assertion failed) !isDynamicShape in function 'parseShape'
> " thrown in the test body.
```

/cc @fengyuentau @vpisarev 

Workaround with disabling this check is ugly as it passes incorrect importing problems to the next layers:

```
DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 243 [ 0 0 0 51 ]
DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 250 [ 0 0 0 34 ]
DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 257 [ 0 0 0 34 ]
DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 264 [ 0 0 0 51 ]
DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 297 [ 0 -4 ]
```

Is it OK?

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -2341,11 +2463,24 @@ void ONNXImporter::parseShape(LayerParams& layerParams, const opencv_onnx::NodeP
...
+    }
+    shapeMat.dims = 1;  // FIXIT Mat 1D
 
+    if (isDynamicShape)
+    {
+        CV_LOG_ERROR(NULL, "DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input " << toString(inpShape, node_proto.input(0)));
+        // FIXIT repair assertion
+        // Disabled to pass face detector tests from #20422
+        // CV_Assert(!isDynamicShape);  // not supported
+    }
```
Comment (alalek) [inline modules/dnn/src/onnx/onnx_importer.cpp:2483]: Error trigger is disabled.

**Comment by alalek** [comment]
:+1:

**Comment by fengyuentau** [comment]
> Face detector (from #20422) was imported through bugs. New checks for unsupported cases emits these errors:
> 
> > DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 243 [ 0 0 0 51 ]
> 
> ```
> [ RUN      ] Objdetect_face_detection.regression
> [ INFO:0@12.252] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (721) populateNet DNN/ONNX: loading ONNX v6 model produced by 'pytorch':1.9. Number of nodes = 111, initializers = 92, inputs = 1, outputs = 3
> [ INFO:0@12.252] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (654) parseOperatorSet DNN/ONNX: ONNX opset version = 9
> [ERROR:0@12.256] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2479) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 243 [ 0 0 0 51 ]
> [ERROR:0@12.256] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (872) handleNode DNN/ONNX: ERROR during processing node with 1 inputs and 1 outputs: [Shape]:(265)
> [ INFO:0@12.256] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (876) handleNode     Input[0] = '243'
> [ INFO:0@12.256] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (880) handleNode     Output[0] = '265'
> unknown file: Failure
> C++ exception with description "OpenCV(4.5.5-pre) /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp:891: error: (-2:Unspecified error) in function 'handleNode'
> > Node [Shape]:(265) parse error: OpenCV(4.5.5-pre) /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp:2480: error: (-215:Assertion failed) !isDynamicShape in function 'parseShape'
> > " thrown in the test body.
> ```
> 
> /cc @fengyuentau @vpisarev
> 
> Workaround with disabling this check is ugly as it passes incorrect importing problems to the next layers:
> 
> ```
> DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 243 [ 0 0 0 51 ]
> DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 250 [ 0 0 0 34 ]
> DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 257 [ 0 0 0 34 ]
> DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 264 [ 0 0 0 51 ]
> DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 297 [ 0 -4 ]
> ```
> 
> Is it OK?

I downloaded the latest code, ran the regression test and did not find this problem. Could you tell me how to reproduce this?

**Pull request merged**